### ISSUE TITLE
Failure rollback doesn't remove symlink for path type repositories

### ISSUE THREAD

**Issue by slbmeh** [composer/composer#5130]
With a composer.json requiring a broken package in a path repository, a subsequent install attempt will fail with a runtime exception:

```
[RuntimeException]
Package package/name cannot install to "/some/path" inside its source at "/some/path"
```

And I expected this to happen:

* Previous failure rollback should unlink the symlink if it was newly created
* Path downloader shouldn't use realpath on the target location for the symlink (realpath for dirname(target) possibly?)

**Comment by curry684** [comment]
> With a composer.json requiring a broken package in a path repository

Can you provide more detailed steps on how to reproduce this? I've seen the path repositories correctly detect and remove broken installs before so that's not the root cause. The "cannot install inside self" error is pretty scary as it implies you're using a pretty risque setup for development. If you provide more info I'll gladly take a look and help determine what's happening and whether it's wrong.

**Comment by slbmeh** [comment]
When setting up a demo it may be isolated to plugins, or any exceptions that occur after install but before installed.json is written.

I created three projects and placed them in a tarball [Project Repositories](http://cdn.stevebuzonas.com/GH-5130.tar.gz)

Overview:
* Create a composer project
* Create a dependency
* Require dependency in project
* Create a plugin dependency that will fail post install e.g. missing requires on `composer-plugin-api`
* Require plugin dependency in project (Will fail leaving symlink)
* Fix plugin dependency
* Require plugin dependency in project (Will fail because realpath on symlink appears as if you are installing into the source target)
* Remove symlink
* Require plugin dependency in project (successful)

**Comment by curry684** [comment]
Hmmm the plugin API isn't my strong suit yet, but from your description above I think this is more likely a flaw in the plugin system than in the path repository, and it just happens to f*ck up harder there as the path repository type is the only one protecting a package from overwriting itself 
(obviously).

@Seldaek can you have a look at this before I spend a lot of time looking for the wrong issue?

**Comment by Seldaek** [comment]
Sounds to me like a bug in the installation of plugins that after an
install if it fails to get enabled we should maybe remove it before
rethrowing the exception. I don't have the code in front of me though.

**Comment by curry684** [comment]
Fixed in #5150.

I've been wondering for a while whether or not to attach the same handler to `update`, but due to the way the installation is ordered we can't do a real rollback there, and at least it will leave the repo in a consistent albeit broken state to fix manually.
